### PR TITLE
[8.x] Avoid SwiftMailer forceReconnection when not running from a queue daemon

### PR DESCRIPTION
**Pull request by alim-shapiev**
I have noticed that emails that are sent in bulk all use separate connections (even though they are not simultaneous, there can be a limit of connections per minute per IP set for an SMTP server). As I've found in multiple pull requests (i.e., [https://github.com/laravel/framework/issues/4573](https://github.com/laravel/framework/issues/4573)), there has been a problem with long lasting SMTP connections due to some problems with SwiftMailer and a simple solution for it:

```
protected function sendSwiftMessage($message)
{
    $this->failedRecipients = [];

    try {
        return $this->swift->send($message, $this->failedRecipients);
    } finally {
        $this->forceReconnection();
    }
}
```

However, these problems are only encountered when a program is executed from a queue daemon (`php artisan queue:work` or `php artisan queue:listen`), so there can be a compromise between dealing with the SSL broken pipe error and sending multiple emails.

I propose setting a flag (or checking somehow else, if there's a better way for it) for when the script is executed from a queue daemon, then, when an email is sent via SMTP and the flag is not present, the SMTP connection will no longer be stopped. This would help in CRON jobs or in case of synchronous mail sending and wouldn't ruin queues a second time.

### Note: this is the second pull request on this matter (first: [https://github.com/laravel/framework/pull/36709](https://github.com/laravel/framework/pull/36709)) due to failing test and consequent reimplementation of the solution.

### CODE REVIEW

File: src/Illuminate/Queue/Worker.php [hunk 1]
```diff
@@ -93,17 +100,20 @@ class Worker
...
      * @param  callable  $isDownForMaintenance
+     * @param  callable  $setQueueDaemonFlag
      * @return void
      */
     public function __construct(QueueManager $manager,
                                 Dispatcher $events,
                                 ExceptionHandler $exceptions,
-                                callable $isDownForMaintenance)
+                                callable $isDownForMaintenance,
+                                callable $setQueueDaemonFlag)
```
Comment (driesvints) [inline src/Illuminate/Queue/Worker.php:110]: This is unfortunately a breaking change. Apps or libraries implementing this class will see their implementation fail be they're not passing the new required dependency.

Comment (anyhon) [reply src/Illuminate/Queue/Worker.php:110]: Could this then be merged to 9.x?

Comment (driesvints) [reply src/Illuminate/Queue/Worker.php:110]: Can't you make it nullable?

Comment (anyhon) [reply src/Illuminate/Queue/Worker.php]: Yes, I don't know why I haven't figured that out myself.

**Comment by anyhon** [comment]
@driesvints sorry for tagging, just wanted to make sure that you've noticed that the compatibility issue is now resolved.

**Comment by driesvints** [comment]
@anyhon thanks. Taylor will review this once he gets to it.

**Comment by morloderex** [comment]
Code looks alright but it might be better to make this figure opt-in?

That way we are not accidentally breaking applications relying on the old behavior.

**Comment by anyhon** [comment]
> Code looks alright but it might be better to make this figure opt-in?
> 
> That way we are not accidentally breaking applications relying on the old behavior.

I want to point out that when the previous fix attempt was implemented, it changed behavior for all mailers (whether it was a simple script that did not really need it or a long-running queue). However, your suggestion is quite reasonable.

Do you think this should be done with a config value or some other way? Probably, something like a MAIL_KEEP_ALIVE = true / false that can be changed in .env, which, in turn, will be sent via to the Mailer instance via its config through something like
```php
$shouldForceReconnection = !($config['keep_alive'] ?? false) || ($this->app['config']['app.is_executed_from_daemon_queue'] ?? false);

$mailer = new Mailer(
    $name,
    $this->app['view'],
    $this->createSwiftMailer($config),
    $this->app['events'],
    $shouldForceReconnection
);
```

**Comment by taylorotwell** [comment]
I'm open to solutions on this problem but don't love this approach.

**Comment by YOzaz** [comment]
This what was proposed on L5 - reset instead of closing, and restarting in case it fails, didn't go through:
```
https://github.com/laravel/framework/pull/8222
```
I've made a package for this a while ago, but it's way behind in time: https://github.com/YOzaz/Laravel-SwiftMailer.

**Pull request closed without merging**